### ISSUE TITLE
Batch rendering not implemented for d3d11

### ISSUE THREAD

**Issue by LauPhi** [libsdl-org/SDL#7534]
OS: Windows 10
SDL: 2.26.4
GCC: 12.2.0 (x86_64-posix-seh-rev2, Built by MinGW-W64 project)

I was using RenderDoc to confirm that the batch rendering was working but it is actually not.  
I am specifying a direct3d11 renderer because RenderDoc can not capture OpenGL 2 frames.  

I am setting the hints like this:
```c
SDL_SetHint(SDL_HINT_RENDER_DRIVER, "direct3d11");
SDL_SetHint(SDL_HINT_RENDER_BATCHING, "1");
```
Then creating the renderer with:
```c
const Uint32 rendererFlags = SDL_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC;
SDL_Renderer* renderer = SDL_CreateRenderer(window, -1, rendererFlags);
```

The texture is rendered using 8 times:
```c
SDL_RenderCopy(renderer, texture, NULL, &imageRect);
```

So no flush should happen until ``SDL_RenderPresent(renderer);`` at the end of the frame.

Here is the frame capture from RenderDoc:
![RenderDoc](https://user-images.githubusercontent.com/14290756/227731169-1d06ffdb-5ac3-4bf1-bb28-ec57899aa5a4.png)

As you can see, there is 8 separate draw calls for essentially the same texture.

I tried spacing the drawn rectangles and it did not enable batching either.

Here is the minimal reproduction project:
[SDLBatchingNotWorking.zip](https://github.com/libsdl-org/SDL/files/11069488/SDLBatchingNotWorking.zip)

**Comment by 1bsyl** [comment]
Indeed ...

d3d11 (nor d3d12) isn't really batching: https://github.com/libsdl-org/SDL/blob/main/src/render/direct3d11/SDL_render_d3d11.c#L2100
gles2 (and gl) does: https://github.com/libsdl-org/SDL/blob/main/src/render/direct3d11/SDL_render_d3d11.c#L2100

There are 3 levels of batching in sdl:
- no batching  / deactivated: each command is sent to the gl driver when the user app emit the command.
- batching (with the hint / by default when no back-end is requested): SDL buffers the user app command, and send them when it has the occasion (flush/present/need to change states).
- command are really batched like gles2. this is back-end specific.

Though you can submit a PR to improve d3d11 ..

**Comment by flibitijibibo** [comment]
Running through SDL_Render issues tagged for 3.0 ABI, sorry if you get multiple e-mails for this...

This could _possibly_ get obsoleted by implementing a backend that uses the GPU API (like the one at #9312), but since SDL_Render targets a _super_ low feature level it would actually be reasonable to keep SDL_render_d3d11, so it may not apply to <D3D11-class hardware...

**Comment by icculus** [comment]
It doesn't look like the Metal or D3D9 renderers try to merge individual draws either.

This is probably worth fixing in all the backends, but it doesn't need to be in the 3.0 ABI milestone. I'm bumping it to 3.2.0, and possibly to 3.x later.

**Comment by python-b5** [comment]
Is this still on the roadmap to be fixed soon? 3.2.0 has come and gone, but D3D11 still does not appear to batch draw calls. Given D3D11 is the default render backend on Windows (at least on my machine), this reduces the default performance of SDL_Renderer by quite a lot. Currently I force a `"gpu"` backend to get around this, but the Vulkan and D3D12 backends both seem to consume more memory than D3D11, which feels a bit wasteful for my app.

**Comment by slouken** [comment]
It turns out this was pretty easy to do, can you try https://github.com/libsdl-org/SDL/pull/14431 and see if that works for you?

**Comment by python-b5** [comment]
Seems to be working well on my end, thanks! RenderDoc shows the draw calls I'd expect to be batched together as batched. Performance on my app now seems comparable to the other backends (though I haven't benchmarked it).

**Comment by slouken** [comment]
> Seems to be working well on my end, thanks! RenderDoc shows the draw calls I'd expect to be batched together as batched. Performance on my app now seems comparable to the other backends (though I haven't benchmarked it).

Thanks for the confirmation!

**Comment by icculus** [comment]
The PR looks okay to me, is there a reason not to merge it at this point? Too large a change for 3.4.0 at this point?

**Comment by slouken** [comment]
> The PR looks okay to me, is there a reason not to merge it at this point? Too large a change for 3.4.0 at this point?

No, it's fine, I'm just doing some final iteration. I'll mark it draft until I'm done.

**Comment by slouken** [comment]
Fixed, thanks!

**Comment by ritalat** [comment]
Looks like D3D9 is still missing batching.

**Comment by slouken** [comment]
> Looks like D3D9 is still missing batching.

Good point, this is fixed in https://github.com/libsdl-org/SDL/commit/5bed8ec60396117e95589128ef8909486112cd84
